firewood/src/merkle.rs

```diff
@@ -355,58 +343,73 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
         val: Vec<u8>,
         root: DiskAddress,
     ) -> Result<(), MerkleError> {
+        let (parents, deleted) = self.insert_and_return_updates(key, val, root)?;
```

Comment:
Had to split this out because of `&self` vs `&mut self` borrowing issues. The alternative was putting the majority of the code in curly braces... I think it's better this way, but open to alternatives.